lib/libzfs/libzfs_pool.c

```diff
@@ -3852,8 +3853,13 @@ zpool_history_unpack(char *buf, uint64_t bytes_read, uint64_t *leftover,
 		/* add record to nvlist array */
 		(*numrecords)++;
 		if (ISP2(*numrecords + 1)) {
-			*records = realloc(*records,
+			tmp = realloc(*records,
 			    *numrecords * 2 * sizeof (nvlist_t *));
+			if (tmp == NULL) {
```

Comment:
I believe we need to decrements `numrecords` in the error case.  This should only be incremented when we've successfully add a new record to `records`.